iree/compiler/Bindings/TFLite/Transforms/WrapEntryPoints.cpp

```diff
@@ -104,27 +502,107 @@ class WrapEntryPointsPass
+      auto inputDynamicDims = std::get<1>(input);
+      SmallVector<Value> dynamicDims;
+      for (auto globalOp : inputDynamicDims.globalOps) {
+        dynamicDims.push_back(entryBuilder.create<IREE::Util::GlobalLoadOp>(
+            arg.getLoc(), globalOp));
+      }
+      callOperands.push_back(entryBuilder.create<IREE::HAL::TensorCastOp>(
+          arg.getLoc(), inputDynamicDims.tensorType, arg, dynamicDims));
+    }
+    auto callOp = entryBuilder.create<mlir::CallOp>(entryFuncOp.getLoc(),
+                                                    entryFuncOp, callOperands);
+    SmallVector<Value> callResults;
+    for (auto output : llvm::zip(callOp.getResults(), outputDynamicDims)) {
+      auto result = std::get<0>(output);
+      auto outputDynamicDims = std::get<1>(output);
+      SmallVector<Value> dynamicDims;
+      for (unsigned i = 0; i < outputDynamicDims.tensorType.getRank(); ++i) {
+        if (outputDynamicDims.tensorType.isDynamicDim(i)) {
+          dynamicDims.push_back(
+              entryBuilder.create<tensor::DimOp>(result.getLoc(), result, i));
+        }
+      }
+      callResults.push_back(entryBuilder.create<IREE::HAL::TensorCastOp>(
+          result.getLoc(), bufferType, result, dynamicDims));
+      for (auto it : llvm::zip(dynamicDims, outputDynamicDims.globalOps)) {
+        auto dynamicDim = std::get<0>(it);
+        auto globalOp = std::get<1>(it);
+        entryBuilder.create<IREE::Util::GlobalStoreOp>(
+            result.getLoc(), dynamicDim, globalOp.getSymbolName());
+      }
+    }
+
+    // We recomputed the shapes of the outputs and can clear the dirty flag.
+    entryBuilder.create<IREE::Util::GlobalStoreOp>(
+        entryFuncOp.getLoc(),
+        entryBuilder.create<arith::ConstantIntOp>(entryFuncOp.getLoc(), 0, 1),
+        dirtyGlobalOp.getSymbolName());
+
+    entryBuilder.create<mlir::ReturnOp>(entryFuncOp.getLoc(), callResults);
+  }
+
+  void wrapEntryPoint(mlir::FuncOp funcOp) {
+    auto loc = funcOp.getLoc();
+    auto namePrefix = ("_tflite_" + funcOp.getName()).str();
+    OpBuilder moduleBuilder(funcOp);
+
+    // Create a variable for each input and output dynamic dim. These variables
+    // may represent fully static shapes - in which case they'll get constant
+    // propagated - or dynamic shapes that will eventually get turned into
+    // dynamic runtime values.
+    auto dynamicDimGlobals =
+        createDynamicDimGlobals(loc, namePrefix, funcOp, moduleBuilder);
 
-    return wrapperFuncOp;
+    // Create internal shape calculation function that updates output shapes if
+    // needed. This is only required if there are dynamic shapes.
+    auto dirtyGlobalOp = moduleBuilder.create<IREE::Util::GlobalOp>(
+        loc, namePrefix + "_shapes_dirty",
+        /*isMutable=*/true, moduleBuilder.getI1Type(),
+        moduleBuilder.getIntegerAttr(moduleBuilder.getI1Type(), 1));
+    dirtyGlobalOp.setPrivate();
+    auto calculateShapeFuncOp = createShapeCalculationFunc(
```

Comment:
I could see us ranting to generalize this soon (ie. For more general conversions to destination passing style).

Reply:
There's a few nice places in the streams pipeline where we can slot that in! I'm excited about getting lit up - it'll also be what's responsible for hoisting allocations of transients so that we can do stream-ordered allocations across function calls 🤤